x-pack/plugins/embeddable_enhanced/public/drilldowns/url_drilldown/url_drilldown_scope.tsx

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/**
+ * This file contains all the logic for mapping from trigger's context and environment context to variables for URL drilldown scope
+ * URL drilldown has 3 sources for variables:
+ *
+ * - Global static variables like, for example, `kibanaUrl`. Such variables won’t change depending on a place where url drilldown is used.
+ * - Context variables are dynamic and different depending on where drilldown is created and used. For example:
+ * - Event variables depend on a trigger context. These variables are dynamically extracted from the action context when drilldown is executed.
+ *
+ * At this point context variables are extracted from `embeddable` and event variables are extracted from trigger's context.
+ * Only 2 triggers are supported at this point: VALUE_CLICK_TRIGGER and RANGE_SELECT_TRIGGER
+ *
+ * In future URL drilldown implementation will provide extension points
+ * for injecting more variables into context and to support more triggers
+ * https://github.com/elastic/kibana/issues/55324
+ *
+ * For now possible variables and logic for extracting those are hardcoded and all contained in this file.
+ */
+
+import React from 'react';
+import {
+  EuiButton,
+  EuiButtonEmpty,
+  EuiModalBody,
+  EuiModalFooter,
+  EuiModalHeader,
+  EuiModalHeaderTitle,
+  EuiRadioGroup,
+} from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n/react';
+import type { Query, Filter, TimeRange } from '../../../../../../src/plugins/data/public';
+import {
+  IEmbeddable,
+  isRangeSelectTriggerContext,
+  isValueClickTriggerContext,
+  RangeSelectContext,
+  ValueClickContext,
+} from '../../../../../../src/plugins/embeddable/public';
+import type { ActionContext, ActionFactoryContext, UrlTrigger } from './url_drilldown';
+import { SELECT_RANGE_TRIGGER } from '../../../../../../src/plugins/ui_actions/public';
+import { OverlayStart } from '../../../../../../src/core/public';
+import { toMountPoint } from '../../../../../../src/plugins/kibana_react/public';
+
+type ContextScopeInput = ActionContext | ActionFactoryContext;
+
+/**
+ * Part of context scope extracted from an embeddable
+ * Expose on the scope as: `{{context.panel.id}}` `{{context.panel.filters.[0]}}`
+ */
+interface EmbeddableUrlDrilldownContextScope {
+  id: string;
+  title?: string;
+  query?: Query;
+  filters?: Filter[];
+  timeRange?: TimeRange;
+  savedObjectId?: string;
+  /**
+   * In case panel supports only 1 index patterns
+   */
+  indexPatternId?: string;
+  /**
+   * In case panel supports more then 1 index patterns
+   */
+  indexPatternIds?: string[];
```

Comment:
Sames as in Dashboard drilldown we will need to figure out how to connect inject/extract/migrate from `PersistentStateService` here, as these are references to saved objects.

Reply:
Fortunately, this is a bit simpler as these are not references to saved object. 
We will store only a template: e.g. `https://go-to-url/{{context.panel.indexPatternId}}`, so in runtime this is going to be substituted with current index pattern id. 

So I guess we would use `PersistentStateService` only for migrating existing templates in case we deprecate one of those fields. But we don't need inject/extract anything. (user could hardcoded some SO ids in template, but I think we can't handle those cases)

Reply:
Good point, now I'm thinking if users will create URL drilldowns for navigation withing Kibana, they will put saved object IDs into the URL. Maybe we should have another variable type, say, "saved object". Where user is presented with saved object picker, and then we know that variable is a reference to some Kibana saved object.